### DISCUSSION TITLE
inspect.has_table() with SQLAlchemy 2.0's MySQLDialect leaves transaction open

### DISCUSSION THREAD

**Discussion by willsthompson** [sqlalchemy/sqlalchemy#10156] [Usage Questions] [answered]
### Describe the bug

After calling `insp.has_table()` using sqlalchemy 1.4, the connection passed to `inspect(connection)` does not have an open transaction, according to `connection.in_transaction()`. However, performing the same steps using sqlalchemy 2 results in an open transaction on the connection following the call to `insp.has_table()`.

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

https://docs.sqlalchemy.org/en/20/core/reflection.html#sqlalchemy.engine.reflection.Inspector.has_table

### SQLAlchemy Version in Use

1.4.49 and 2.0.19

### DBAPI (i.e. the database driver)

mysqlclient

### Database Vendor and Major Version

MySQL 8

### Python Version

3.10

### Operating system

macOS

### To Reproduce

```python
# Run using `sqlalchemy==1.4.49`, then `sqlalchemy==2.0.49`

from sqlalchemy import create_engine
from sqlalchemy import inspect

engine = create_engine(url="mysql+mysqldb://root:password@127.0.0.1:3306/")
with engine.connect() as conn:
    insp = inspect(conn)
    ret = insp.has_table("missing", "default")
    assert conn.in_transaction() is False
```

### Error

The issue does not produce an error, only a behavior change.

### Additional context

I can't be sure this isn't user error, but the documentation doesn't specify here and the behavior change broke something of ours dependent on the 1.4.x behavior. We're in the process of migrating to 2.0, and if it's not a bug, I was hoping to at least get clarification on why/how best to handle this change.

I noticed the function changed significantly across major versions, from https://github.com/sqlalchemy/sqlalchemy/blob/a7438bba206f5ffa2a14cd9398354ba78fed7b6e/lib/sqlalchemy/dialects/mysql/base.py#L2640-L2660 to https://github.com/sqlalchemy/sqlalchemy/blob/9e89d2cbbf7ccef05579472b94c90124a1ecf9e3/lib/sqlalchemy/dialects/mysql/base.py#L2684-L2727

Thank you!

**Comment by zzzeek** [comment, marked as answer]
hi there -

Upgrading an existing application from 1.4 to 2.0 should start with following the [migration guide](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#).   Step four in this guide is to keep the program running under SQLAlchemy 1.4, [adding the parameter `future=True` to `create_engine()`](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#migration-to-2-0-step-four-use-the-future-flag-on-engine) calls as follows - this allows 2.0 style engine behaviors to take effect while still running SQLAlchemy 1.4:

```py
engine = create_engine(url="mysql+mysqldb://scott:tiger@localhost/test", future=True)
with engine.connect() as conn:
    insp = inspect(conn)
    ret = insp.has_table("missing", "default")
    assert conn.in_transaction() is False
```

when we make this change, note that the program given has the identical behavior in 1.4 as it does in 2.0.   "future" mode allows you to test your program under 1.4 and observe up front what major behaviors will affect it.   

in this case the behavior you are seeing is probably the most major change in the 1.4-> 2.0 series which is [library level autocommit is removed](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#library-level-but-not-driver-level-autocommit-removed-from-both-core-and-orm).  SQLAlchemy no longer automatically ends the transaction that is necessarily begun when SQL commands are emitted on a connection.

to end the transaction after an operation, using a `begin()` block is most expedient:

```py
engine = create_engine(url="mysql+mysqldb://scott:tiger@localhost/test", future=True)
with engine.connect() as conn:
    with conn.begin():
        insp = inspect(conn)
        ret = insp.has_table("missing", "default")
    assert conn.in_transaction() is False
```

hope this helps!

**Reply by gordthompson** [reply]
To flesh out the behaviours:

Scenario 1: inspect(engine)

```python
import sqlalchemy as sa

engine = sa.create_engine("mysql://scott:tiger@localhost:3307/mydb")

insp = sa.inspect(engine)
insp.has_table("foo")
"""From MySQL general log:
1 Query	DESCRIBE `mydb`.`foo`
1 Query	ROLLBACK
"""
```

Scenario 2: context manager connection

```python
with engine.connect() as conn:
    insp = sa.inspect(conn)
    insp.has_table("foo")
    """From MySQL general log:
    2 Query	DESCRIBE `mydb`.`foo`    
    """
"""From MySQL general log:
2 Query	ROLLBACK
"""
print("Done.")
# (no log activity)
```

Scenario 3: nested context manager for conn.begin()

```python
with engine.connect() as conn:
    with conn.begin():
        insp = sa.inspect(conn)
        insp.has_table("foo")
        """From MySQL general log:
        3 Query	DESCRIBE `mydb`.`foo`
        """
     """From MySQL general log:
     3 Query	COMMIT
     """
    print("exit begin()")
    # (no log activity)
print("Done.")
# (no log activity)
```

**Reply by willsthompson** [reply]
Thank you both for the detailed explanations and examples!
